### PR TITLE
Implement avatar miniatures

### PR DESCRIPTION
**Pull request by kontrollanten**
## Description
Generate avatar miniatures to optimize bandwidth when showing smaller avatars.

## Related issues
#4560 

## Has this been tested?

- [x] Yes

### CODE REVIEW

**Comment by kontrollanten** [comment]
@Chocobozzz How should we handle the activity pub part here? As I understand the account/channel avatar is only an object (icon). Shall I migrate it to an array instead?

**Comment by Chocobozzz** [comment]
> How should we handle the activity pub part here? As I understand the account/channel avatar is only an object (icon). Shall I migrate it to an array instead?

Unfortunately it seems we don't support `.icon` array: https://github.com/Chocobozzz/PeerTube/blob/develop/server/lib/activitypub/actors/shared/object-to-model-attributes.ts#L38

So you'll have to handle `.icon` arrays **and** rebuild the avatar miniature to not break federation, and in a few versions later we'll send an array.

**Comment by kontrollanten** [comment]
@Chocobozzz I didn't add any tests, but maybe it's enough that I just updated the existing tests? It would be good to add a test to verify that the icons array works in activity pub, but I couldn't figure out how to do that.

**Comment by kontrollanten** [comment]
@Chocobozzz Are you waiting for anything from me to be able to merge this?

**Comment by Chocobozzz** [comment]
> @Chocobozzz Are you waiting for anything from me to be able to merge this?

Sorry, I forgot to review this PR.

I think we should instead have an `.avatars` object that is an array. Then we can have `width`/`height` properties so the client choose the more appropriate one

**Comment by kontrollanten** [comment]
That sounds good. Do you think we should just change the API response, or shall I remove the whole "avatar miniature" from the backend? Maybe it's better to refactor the whole way, and just treat all avatars the same.

**Comment by Chocobozzz** [comment]
I think we should remove the avatar miniature concept from backend

**Comment by kontrollanten** [comment]
Holy Moly. This was a heavy one, but finally I think I got everything working. Some comments:

* Maybe we should add a test to verify that activity pub responses with icon as array is working?
* In some cases `findOne` only returned an array with a single avatar, therefore I changed them to `findAll`.
* There was some issues with `findAndCountAll` which I solved by making them distinct. https://github.com/sequelize/sequelize/issues/9481
* I'm not used to work with sequelize so I'm not sure if I made it by the book.

**Comment by Chocobozzz** [comment]
I'm reviewing/updating this PR (could take some time)

**Review by Chocobozzz — COMMENTED** [review]
* I'll update search index to include `avatars`/`banners` attribute so we don't have to normalize the search index responses
 * I'll add previous `avatar`/`banner` attributes to not break REST API
 * I'll optimize the migration, I'm afraid it's too slow with instances with big federation 

I think I'll also remove small avatar generation from requests from AP. We'll only have 1 big avatar in the array, but I don't think it's a bigger issue than overloading the server by generating small avatars

File: server/models/actor/actor-follow.ts [hunk 1]
```diff
@@ -498,7 +498,7 @@ export class ActorFollowModel extends Model<Partial<AttributesOnly<ActorFollowMo
     }
 
     const query = {
-      attributes: [],
+      attributes: SORTABLE_COLUMNS.USER_SUBSCRIPTIONS, // Alt use this method instead https://github.com/sequelize/sequelize/issues/4553
```
[the remark this answers is not in the document]
Comment (kontrollanten) [reply server/models/actor/actor-follow.ts:501]: Without including those attributes, it wasn't possible to sort by those columns

File: server/lib/activitypub/process/process-update.ts [hunk 2]
```diff
@@ -99,8 +99,6 @@ async function processUpdateCacheFile (byActor: MActorSignature, activity: Activ
 async function processUpdateActor (actor: MActorFull, activity: ActivityUpdate) {
   const actorObject = activity.object as ActivityPubActor
 
-  logger.debug('Updating remote account "%s".', actorObject.url)
```
[the remark this answers is not in the document]
Comment (kontrollanten) [reply server/lib/activitypub/process/process-update.ts:102]: Oh, that was a misstake, I thought it was mine own debugging output

File: scripts/migrations/peertube-4.2.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  console.log('Generation finished!')
+}
+
+async function generateSmallerAvatarIfNeeded (accountOrChannel: MAccountDefault | MChannelDefault) {
+  if (accountOrChannel.Actor.Avatars.length !== 1) {
+    return
+  }
+
+  await generateSmallerAvatar(accountOrChannel.Actor, (t: Transaction, updateActor: MActorImages) => {
+    accountOrChannel.Actor = Object.assign(updateActor, { Server: accountOrChannel.Actor.Server })
```
[the remark this answers is not in the document]
Comment (kontrollanten) [reply scripts/migrations/peertube-4.2.ts:64]: I actually don't know. If it's null it doesn't seem necessary to keep it.

[the remark this answers is not in the document]
Comment (kontrollanten) [reply scripts/migrations/peertube-4.2.ts:64]: Now I found out why. It was due a TypeScript error; 

```
Type 'MActorImages' is not assignable to type 'MActorDefault'.
  Property 'Server' is missing in type 'MActorImages' but required in type 'Use<"Server", MServer>'.
 ```

But I guess that can be resolved by just changing the type explicitly.

**Comment by Chocobozzz** [comment]
I added in the tests the case where accounts and channels have avatars. Most of the test failed because of broken SQL queries generated by sequelize due to the fact that now Avatars is a multi association (invalid counts, incorrect subquery etc). I spent days to fix and rewrite some of them. There's still room for refactoring but I need to move on so it will be enough for this PR.

**Pull request merged**